File: frontend/src/lib/api/pipeline.ts
import { apiRequest, type ApiRequestOptions } from "@/lib/api/request";

export type PipelineNodeStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "skipped";

export type PipelineNode = {
  id: string;
  label: string;
  section: "ingestion" | "query" | "evaluation";
  description?: string;
  status: PipelineNodeStatus;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
  metrics?: Record<string, unknown>;
};

export type PipelineEdge = {
  id: string;
  source: string;
  target: string;
};

export type PipelineRunGraphResponse = {
  pipeline_run_id: string;
  pipeline_type: string;
  status: string;
  nodes: PipelineNode[];
  edges: PipelineEdge[];
};

export type PipelineRunResolveResponse = {
  pipeline_run_id: string;
  pipeline_type: string;
  status: string;
};

export type PipelineNodeDetailResponse = {
  node_id: string;
  title: string;
  description: string;
  status: PipelineNodeStatus;
  inputs: Record<string, unknown>;
  outputs: Record<string, unknown>;
  config: Record<string, unknown>;
  logs: string[];
  error_message: string | null;
  error_details: Record<string, unknown>;
  metrics: Record<string, unknown>;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
};

export type PipelineStepListResponse = {
  steps: string[];
};

export type PipelineRequestOptions = Pick<
  ApiRequestOptions,
  "apiBaseUrl" | "token" | "organizationId" | "signal"
>;

export async function fetchPipelineSteps(
  options: PipelineRequestOptions = {},
): Promise<PipelineStepListResponse> {
  return apiRequest<PipelineStepListResponse>("/pipeline/steps", {
    ...options,
    retry: { maxRetries: 1 },
  });
}

export async function fetchPipelineRunGraph(
  runId: string,
  options: PipelineRequestOptions = {},
): Promise<PipelineRunGraphResponse> {
  return apiRequest<PipelineRunGraphResponse>(
    `/pipeline/runs/${encodeURIComponent(runId)}`,
    options,
  );
}

export async function resolvePipelineRun(
  params: {
    run_type?: string | null;
    document_id?: string | null;
    chat_message_id?: string | null;
    evaluation_run_id?: string | null;
  },
  options: PipelineRequestOptions = {},
): Promise<PipelineRunResolveResponse> {
  return apiRequest<PipelineRunResolveResponse>("/pipeline/runs/resolve", {
    ...options,
    query: {
      run_type: params.run_type ?? undefined,
      document_id: params.document_id ?? undefined,
      chat_message_id: params.chat_message_id ?? undefined,
      evaluation_run_id: params.evaluation_run_id ?? undefined,
    },
  });
}

export async function fetchPipelineNodeDetail(
  runId: string,
  nodeId: string,
  options: PipelineRequestOptions = {},
): Promise<PipelineNodeDetailResponse> {
  return apiRequest<PipelineNodeDetailResponse>(
    `/pipeline/runs/${encodeURIComponent(runId)}/nodes/${encodeURIComponent(nodeId)}`,
    options,
  );
}

File: backend/app/domains/pipeline/schemas/pipeline.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class PipelineStepListResponse(BaseModel):
    steps: list[str]


class PipelineNodeResponse(BaseModel):
    id: str
    label: str
    section: Literal["ingestion", "query", "evaluation"]
    description: str | None = None
    status: Literal["pending", "running", "completed", "failed", "skipped"]
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    metrics: dict[str, object] = Field(default_factory=dict)


class PipelineEdgeResponse(BaseModel):
    id: str
    source: str
    target: str


class PipelineRunGraphResponse(BaseModel):
    pipeline_run_id: str
    pipeline_type: str
    status: str
    nodes: list[PipelineNodeResponse]
    edges: list[PipelineEdgeResponse]


class PipelineRunResolveResponse(BaseModel):
    pipeline_run_id: str
    pipeline_type: str
    status: str


class PipelineNodeDetailResponse(BaseModel):
    node_id: str
    title: str
    description: str
    status: Literal["pending", "running", "completed", "failed", "skipped"]
    inputs: dict[str, object] = Field(default_factory=dict)
    outputs: dict[str, object] = Field(default_factory=dict)
    config: dict[str, object] = Field(default_factory=dict)
    logs: list[str] = Field(default_factory=list)
    error_message: str | None = None
    error_details: dict[str, object] = Field(default_factory=dict)
    metrics: dict[str, object] = Field(default_factory=dict)
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
